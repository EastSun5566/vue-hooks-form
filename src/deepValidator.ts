import Validator, { RuleItem, Rules, ValidateSource } from 'async-validator'
import {
  toPath, setWith, get, set,
} from './utils'

const getRulePath = (path: any) => toPath(path).join('.fields.')

const getRule = (rules: Rules, path: any) => {
  const rulePath = getRulePath(path)
  return get(rules, rulePath) as RuleItem
}

const setRule = (rules: Rules, path: any, rule: RuleItem) => {
  const rulePath = getRulePath(path)
  return setWith(rules, rulePath, rule, (pathValue, key) => {
    if (key !== 'fields') {
      return pathValue || ({ type: 'object' } as RuleItem)
    }
    return pathValue
  })
}
const DeepValidator = (rules: Rules = {}) => {
  const registerRule = (path: any, rule: RuleItem) => {
    setRule(rules, path, rule)
  }
  const validate = async (data: ValidateSource) => {
    try {
      return await new Validator(rules).validate(data)
    } catch ({ errors, fields }) {
      throw fields
    }
  }
  return {
    getRules: () => rules,
    registerRule,
    validate,
    validateField: async (path: any, value: any) => {
      const fieldRule = setRule({}, path, getRule(rules, path))
      try {
        return await new Validator(fieldRule).validate(
          set({}, path, value),
        )
      } catch ({ fields }) {
        throw fields[path]
      }
    },
  }
}

export default DeepValidator
