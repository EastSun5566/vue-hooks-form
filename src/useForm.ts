import {
  reactive, computed, ref, Ref,
} from 'vue'
import { RuleItem } from 'async-validator'
import DeepValidator from './DeepValidator'
import {
  isAllUnmounted, get, set, toPathString,
} from './utils'

export type FormOptions<Values extends object> = {
  defaultValues?: Values;
  shouldUnregister?: boolean;
}

export type FieldOptions = {
  rule?: RuleItem;
}
export type Error = {
  message: string;
  field: string;
}

export type ErrorFields = {
  [field: string]: Error[] | undefined;
}

export const useForm = <T extends object>({
  defaultValues = {} as T,
  shouldUnregister = true,
}: FormOptions<T>) => {
  const validator = DeepValidator({})
  const fieldsRef = ref<{ [key: string]: Set<Ref<HTMLElement | null>> }>({})
  const fieldValues = reactive(defaultValues) as any

  const errorFields = reactive({} as ErrorFields)

  // make errorFields is reactive
  const clearErrorFields = () => {
    Object.keys(errorFields).forEach((key) => {
      delete errorFields[key]
    })
  }
  const setErrorFields = (errors: ErrorFields) => {
    Object.keys(errors).forEach((key) => {
      errorFields[key] = errors[key]
    })
  }
  const useField = (path: string | (string | number)[], options: FieldOptions = {}) => {
    const pathStr = toPathString(path)
    const fieldRef = ref<HTMLElement | null>(null)
    const { rule } = options
    if (rule) {
      validator.registerRule(pathStr, rule)
    }
    const value = computed({
      get: () => get(fieldValues, pathStr),
      set: (newValue) => {
        set(fieldValues, pathStr, newValue)
      },
    })
    const getRef = (nodeRef: HTMLElement) => {
      fieldRef.value = nodeRef
      const nodeSet = fieldsRef.value[pathStr] || new Set()
      nodeSet.add(fieldRef)
      fieldsRef.value[pathStr] = nodeSet
    }
    return reactive({
      ref: getRef,
      value,
      errors: computed(() => errorFields[pathStr]),
    })
  }
  const getFieldValues = () => Object.keys(fieldsRef.value).reduce((acc, path) => {
    // only return fields that exit on page
    const value = get(fieldValues, path)
    if (!shouldUnregister) {
      set(acc, path, value)
      return acc
    }
    if (!isAllUnmounted(fieldsRef.value[path])) {
      set(acc, path, value)
      return acc
    }
    return acc
  }, {})
  const validateFields = async () => {
    try {
      await validator.validate(getFieldValues())
      clearErrorFields()
    } catch (error) {
      setErrorFields(error)
      throw error
    }
  }

  const validateField = async (path: any) => {
    try {
      await validator.validateField(path, get(fieldValues, path))
      delete errorFields[path]
    } catch (error) {
      errorFields[path] = error
      throw error
    }
  }
  return reactive({
    values: fieldValues as T,
    useField,
    get: (path: string, defaultValue: any) => {
      get(fieldValues, path, defaultValue)
    },
    set: (path: string, value: any) => {
      set(fieldValues, path, value)
    },
    getFieldValues,
    validateFields,
    validateField,
    errorFields,
  })
}
