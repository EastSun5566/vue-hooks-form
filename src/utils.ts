import { Ref } from 'vue'

export const isUnmounted = (fieldRef: Ref) => fieldRef.value === null

export const isAllUnmounted = (fieldRefs?: Set<Ref>) => {
  if (fieldRefs === undefined) {
    return true
  }
  return [...fieldRefs].every(isUnmounted)
}
