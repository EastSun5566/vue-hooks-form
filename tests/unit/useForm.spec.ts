import { renderHook, waitForNextUpdate } from '../utils'
import { useForm } from '../../src'

describe('useFrom', () => {
  test('basic', async (done) => {
    const defaultValues = {
      name: 'wang',
      age: 1,
      info: {
        email: 'hello@example.com',
        city: 'beijing',
      },
      notInclude: 'notInclude',
    }
    const { result } = renderHook(() => useForm({
      defaultValues,
    }))
    if (result !== undefined) {
      const nameField = result.useField('name', {
        rule: { required: true },
      })
      expect(result.values).toEqual(defaultValues)
      expect(result.errors).toEqual({})
      expect(nameField.value).toBe('wang')
      nameField.value = 'victor'
      expect(nameField.value).toBe('victor')
      await nameField.validate()
      expect(result.errors.name).toBe(undefined)
      nameField.value = undefined
      await waitForNextUpdate()
      expect(nameField.error).toBeDefined()
      expect(result.errors.name).toBeDefined()
      done()
    } else {
      throw Error
    }
  })
  test('nested', async (done) => {
    const { result } = renderHook(() => useForm())
    if (result !== undefined) {
      expect(result.values).toEqual({})
      const nestedField = result.useField('a.b.c.d', {
        rule: { required: true },
      })
      expect(nestedField.value).toBeUndefined()
      expect(nestedField.error).toBeUndefined()
      nestedField.value = 1
      expect(result.get('a.b.c.d')).toBe(1)
      await waitForNextUpdate()
      expect(nestedField.error).toBeDefined()
    } else {
      throw Error
    }
    done()
  })
})
