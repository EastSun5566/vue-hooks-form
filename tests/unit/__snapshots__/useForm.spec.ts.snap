// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useFrom renders props.msg when passed 1`] = `
VueWrapper {
  "__app": Object {
    "_component": Object {
      "__emits": Object {
        "0": Object {},
      },
      "__props": Object {
        "0": Array [
          Object {},
          Array [],
        ],
      },
      "name": "VTU_ROOT",
      "render": [Function],
    },
    "_container": <div
      data-v-app=""
      id="app"
    >
      <div>
        <input
          id="name"
          type="text"
        />
        <input
          type="text"
        />
        <input
          type="email"
        />
        <button>
           submit 
        </button>
      </div>
    </div>,
    "_context": Object {
      "app": [Circular],
      "components": Object {},
      "config": Object {
        "errorHandler": undefined,
        "globalProperties": Object {},
        "isCustomElement": [Function],
        "isNativeTag": [Function],
        "optionMergeStrategies": Object {},
        "performance": false,
        "warnHandler": undefined,
      },
      "directives": Object {},
      "mixins": Array [
        Object {
          "__emits": Object {
            "0": null,
          },
          "__props": Object {
            "0": Array [],
          },
          "beforeCreate": [Function],
        },
        Object {
          "__emits": Object {
            "0": null,
          },
          "__props": Object {
            "0": Array [],
          },
          "beforeCreate": [Function],
        },
      ],
      "provides": Object {},
      "reload": [Function],
    },
    "_props": null,
    "_uid": 0,
    "component": [Function],
    "config": Object {
      "errorHandler": undefined,
      "globalProperties": Object {},
      "isCustomElement": [Function],
      "isNativeTag": [Function],
      "optionMergeStrategies": Object {},
      "performance": false,
      "warnHandler": undefined,
    },
    "directive": [Function],
    "mixin": [Function],
    "mount": [Function],
    "provide": [Function],
    "unmount": [Function],
    "use": [Function],
    "version": "3.0.0",
  },
  "__setProps": [Function],
  "componentVM": Object {
    "__emitted": Object {},
    "ageField": Object {
      "errors": undefined,
      "ref": [Function],
      "value": 1,
    },
    "defaultValues": Object {
      "age": 1,
      "info": Object {
        "city": "beijing",
        "email": "hello@example.com",
      },
      "name": "wang",
      "notInclude": "notInclude",
    },
    "emailField": Object {
      "errors": undefined,
      "ref": [Function],
      "value": "hello@example.com",
    },
    "getFieldValues": [Function],
    "nameField": Object {
      "errors": undefined,
      "ref": [Function],
      "value": "wang",
    },
    "state": Object {
      "visible": true,
    },
    "submit": [Function],
    "validateField": [Function],
    "validateFields": [Function],
    "values": Object {
      "age": 1,
      "info": Object {
        "city": "beijing",
        "email": "hello@example.com",
      },
      "name": "wang",
      "notInclude": "notInclude",
    },
  },
  "rootVM": Object {
    "__emitted": Object {},
  },
}
`;
